packages/lsd-react/src/components/ButtonGroup/ButtonGroup.tsx

```diff
@@ -0,0 +1,51 @@
+import clsx from 'clsx'
+import React from 'react'
+import {
+  CommonProps,
+  omitCommonProps,
+  useCommonProps,
+} from '../../utils/useCommonProps'
+import { buttonGroupClasses } from './ButtonGroup.classes'
+import { Button } from '../Button/Button'
+
+export type ButtonGroupProps = CommonProps &
+  React.HTMLAttributes<HTMLDivElement> & {
+    variant?: 'outlined' | 'filled'
+    size?: 'large' | 'medium' | 'small'
+    disabled?: boolean
+  }
+
+export const ButtonGroup: React.FC<ButtonGroupProps> & {
+  classes: typeof buttonGroupClasses
+} = ({ size = 'medium', variant = 'outlined', children, ...props }) => {
+  const commonProps = useCommonProps(props)
+
+  return (
+    <div
+      {...omitCommonProps(props)}
+      className={clsx(
+        commonProps.className,
+        props.className,
+        buttonGroupClasses.root,
+      )}
+    >
+      {React.Children.map(children, (child) => {
+        // Check if the child is a valid React element and if it's of type Button.
+        if (React.isValidElement(child) && child.type === Button) {
+          // Clone the child element (Button) and provide the updated props.
+          return React.cloneElement(child, {
+            ...child.props,
+            size,
+            variant,
+            disabled: props.disabled || child.props.disabled,
+          })
+        }
+
+        // If not a Button, return the child unchanged.
+        return child
+      })}
```

Comment:
Thanks for the comments. Regarding this point:

>  won't work in case we need to wrap a button inside another element, like an anchor. I'd suggest using a context instead.

I think this goes a little bit outside of what was asked. The ButtonGroup should be used for buttons - that's what's on the designs. I added the following:

`        // If not a Button, return the child unchanged.
        return child
`

because it was easy to do, and in my opinion better than the alternatives: return null, or perhaps throw and error.

However, if you feel strongly about the context solution - could you please elaborate a bit more? What would the context provide?

Regarding your other point:

> Also, the child component properties should override the group's, not the other way around.

This could also create situations that differ from the designs - for example, a button group where buttons have different sizes / variants (which is not in the designs). I only added the disable prop like so: `disabled: props.disabled || child.props.disabled,` because I feel like the disabled prop touches more on the functionality side of things.

> The common props should be passed to the child components as well

Agreed.

Reply:
> A button wrapped inside an anchor element is still a button.

I'll have to disagree here. The name of this component is ButtonGroup - it's not AnchorWithButtonChildGroup. It's supposed to group multiple button components - not anchor components, or divs, or whatever else.

> And, what's on the designs does not account for all use cases (...) if there's anything on Figma suggesting how LSD should be used, it should be seen more as a guideline than hard-written rules. 

I'm sorry but this is not what I've been told. I've been told to follow the figma designs as closely as possible. Also if I start thinking like this, then this component can be anything really - a button group, a div group, a paragraph group... when does it stop? It is not clear, and unfair in my opinion to expect me (or anyone really) to know all of this beforehand.

> Please take a look at the IconButtonGroup and IconButton and follow the same approach for this.

This is the type of stuff I expect to be told before starting to work on a component - not after I've spent multiple hours implementing it in a specific way. This will mean rewriting the whole thing, which as you can imagine is not ideal given we only have a couple of days left until the end of the sprint.

Reply:
> But you need to understand our designers are not web developers; we as the developers of this LSD implementation must try to make a better developer experience.

A better developer experience doesn't mean allow users to pass in any possible component as child, and expect the parent to behave correctly. And, most importantly, a "better developer experience" certainly does not mean "implement this component exactly like the icon group" - which is what you're asking me to do.

> We have to ensure the patterns and coding style are 100% consistent across all components; same goes for the functionality of similar components. You must understand, this is what this project requires. 

Again, as I've said in my previous response, this is the type of stuff I expect to be told before starting to work on a component. You can't possibly expect me or anyone to magically guess that you want this component to be implemented like the IconGroupComponent - or any other component (there are quite a few).

As a sidenote - you've said twice now the I should  "use context". However, the meaning of this is not clear - context can be used in many different ways. You have to be more specific. If I didn't ask you what you meant in my last comment, I would have never guessed how you wanted this to be implemented.

> Besides all these, I see using React.cloneElement([which is also not recommended to use anymore] (https://react.dev/reference/react/legacy)) as an anti-pattern where you could use a context.

Did you read the link you posted? It says JSX is recomment over createElement. That's it. Nowhere does it say it's an anti pattern. [In fact, if you follow the link to createElement, they clearly state that both using createElement and JSX are both valid approaches](https://react.dev/reference/react/createElement).

> If you don’t have the time, or simply don’t feel happy to adjust these changes for any reason, please let me know and I’d do it myself

I will not be implementing these changes. And a talk with @amirhouieh is in order - because if every time I submit a PR I have to magically guess how you want things to be coded, this is not going to work.